File: binding/include/libqfile.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qntfspermissioncheckguard.html
pub const qntfspermissioncheckguard = struct {

    /// New constructs a new QNtfsPermissionCheckGuard object.
    pub fn New() QtC.QNtfsPermissionCheckGuard {
        return qtc.QNtfsPermissionCheckGuard_new();
    }


    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QNtfsPermissionCheckGuard_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qfile.html
pub const qfile = struct {

    /// New constructs a new QFile object.
    pub fn New() QtC.QFile {
        return qtc.QFile_new();
    }


    /// New2 constructs a new QFile object.
    pub fn New2(name: []const u8) QtC.QFile {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QFile_new2(name_str);
    }


    /// New3 constructs a new QFile object.
    pub fn New3(parent: ?*anyopaque) QtC.QFile {
        return qtc.QFile_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QFile object.
    pub fn New4(name: []const u8, parent: ?*anyopaque) QtC.QFile {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QFile_new4(name_str, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_FileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnFileName(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QFile_OnFileName(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_QBaseFileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FilesystemFileName(self: ?*anyopaque, ) std::filesystem::path {
        return qtc.QFile_FilesystemFileName(@ptrCast(self));
    }

    pub fn SetFileName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QFile_SetFileName(@ptrCast(self), name_str);
    }

    pub fn EncodeName(fileName: []const u8, allocator: std.mem.Allocator) []u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const _bytearray: qtc.libqt_string = qtc.QFile_EncodeName(fileName_str);
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qfile.EncodeName: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn DecodeName(localFileName: []u8, allocator: std.mem.Allocator) []const u8 {
        const localFileName_str = qtc.libqt_string{
    .len = localFileName.len,
    .data = localFileName.ptr,
};
const _str = qtc.QFile_DecodeName(localFileName_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.DecodeName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn DecodeName2(localFileName: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_DecodeName2(@ptrCast(localFileName));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.DecodeName2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Exists(self: ?*anyopaque, ) bool {
        return qtc.QFile_Exists(@ptrCast(self));
    }

    pub fn Exists2(fileName: []const u8) bool {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QFile_Exists2(fileName_str);
    }

    pub fn SymLinkTarget(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_SymLinkTarget(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.SymLinkTarget: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SymLinkTarget2(fileName: []const u8, allocator: std.mem.Allocator) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const _str = qtc.QFile_SymLinkTarget2(fileName_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.SymLinkTarget2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FilesystemSymLinkTarget(self: ?*anyopaque, ) std::filesystem::path {
        return qtc.QFile_FilesystemSymLinkTarget(@ptrCast(self));
    }

    pub fn Remove(self: ?*anyopaque, ) bool {
        return qtc.QFile_Remove(@ptrCast(self));
    }

    pub fn Remove2(fileName: []const u8) bool {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QFile_Remove2(fileName_str);
    }

    pub fn MoveToTrash(self: ?*anyopaque, ) bool {
        return qtc.QFile_MoveToTrash(@ptrCast(self));
    }

    pub fn MoveToTrash2(fileName: []const u8) bool {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QFile_MoveToTrash2(fileName_str);
    }

    pub fn Rename(self: ?*anyopaque, newName: []const u8) bool {
        const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QFile_Rename(@ptrCast(self), newName_str);
    }

    pub fn Rename2(oldName: []const u8, newName: []const u8) bool {
        const oldName_str = qtc.libqt_string{
    .len = oldName.len,
    .data = oldName.ptr,
};
const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QFile_Rename2(oldName_str, newName_str);
    }

    pub fn Link(self: ?*anyopaque, newName: []const u8) bool {
        const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QFile_Link(@ptrCast(self), newName_str);
    }

    pub fn Link2(fileName: []const u8, newName: []const u8) bool {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QFile_Link2(fileName_str, newName_str);
    }

    pub fn Copy(self: ?*anyopaque, newName: []const u8) bool {
        const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QFile_Copy(@ptrCast(self), newName_str);
    }

    pub fn Copy2(fileName: []const u8, newName: []const u8) bool {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QFile_Copy2(fileName_str, newName_str);
    }

    pub fn Open(self: ?*anyopaque, flags: i32) bool {
        return qtc.QFile_Open(@ptrCast(self), @intCast(flags));
    }

    /// Allows for overriding the related default method
    pub fn OnOpen(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) bool) void {
        qtc.QFile_OnOpen(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseOpen(self: ?*anyopaque, flags: i32) bool {
        return qtc.QFile_QBaseOpen(@ptrCast(self), @intCast(flags));
    }

    pub fn Open2(self: ?*anyopaque, flags: i32, permissions: i32) bool {
        return qtc.QFile_Open2(@ptrCast(self), @intCast(flags), @intCast(permissions));
    }

    pub fn Open3(self: ?*anyopaque, f: *FILE, ioFlags: i32) bool {
        return qtc.QFile_Open3(@ptrCast(self), @ptrCast(f), @intCast(ioFlags));
    }

    pub fn Open4(self: ?*anyopaque, fd: i32, ioFlags: i32) bool {
        return qtc.QFile_Open4(@ptrCast(self), fd, @intCast(ioFlags));
    }

    pub fn Size(self: ?*anyopaque, ) i64 {
        return qtc.QFile_Size(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSize(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i64) void {
        qtc.QFile_OnSize(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSize(self: ?*anyopaque, ) i64 {
        return qtc.QFile_QBaseSize(@ptrCast(self));
    }

    pub fn Resize(self: ?*anyopaque, sz: i64) bool {
        return qtc.QFile_Resize(@ptrCast(self), sz);
    }

    /// Allows for overriding the related default method
    pub fn OnResize(self: ?*anyopaque, callback: *const fn (?*anyopaque, i64) callconv(.c) bool) void {
        qtc.QFile_OnResize(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseResize(self: ?*anyopaque, sz: i64) bool {
        return qtc.QFile_QBaseResize(@ptrCast(self), sz);
    }

    pub fn Resize2(filename: []const u8, sz: i64) bool {
        const filename_str = qtc.libqt_string{
    .len = filename.len,
    .data = filename.ptr,
};
return qtc.QFile_Resize2(filename_str, sz);
    }

    pub fn Permissions(self: ?*anyopaque, ) i32 {
        return qtc.QFile_Permissions(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnPermissions(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QFile_OnPermissions(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePermissions(self: ?*anyopaque, ) i32 {
        return qtc.QFile_QBasePermissions(@ptrCast(self));
    }

    pub fn Permissions2(filename: []const u8) i32 {
        const filename_str = qtc.libqt_string{
    .len = filename.len,
    .data = filename.ptr,
};
return qtc.QFile_Permissions2(filename_str);
    }

    pub fn SetPermissions(self: ?*anyopaque, permissionSpec: i32) bool {
        return qtc.QFile_SetPermissions(@ptrCast(self), @intCast(permissionSpec));
    }

    /// Allows for overriding the related default method
    pub fn OnSetPermissions(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) bool) void {
        qtc.QFile_OnSetPermissions(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetPermissions(self: ?*anyopaque, permissionSpec: i32) bool {
        return qtc.QFile_QBaseSetPermissions(@ptrCast(self), @intCast(permissionSpec));
    }

    pub fn SetPermissions2(filename: []const u8, permissionSpec: i32) bool {
        const filename_str = qtc.libqt_string{
    .len = filename.len,
    .data = filename.ptr,
};
return qtc.QFile_SetPermissions2(filename_str, @intCast(permissionSpec));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFile_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfile.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn MoveToTrash22(fileName: []const u8, pathInTrash: []const u8) bool {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const pathInTrash_str = qtc.libqt_string{
    .len = pathInTrash.len,
    .data = pathInTrash.ptr,
};
return qtc.QFile_MoveToTrash22(fileName_str, pathInTrash_str);
    }

    pub fn Open32(self: ?*anyopaque, f: *FILE, ioFlags: i32, handleFlags: i32) bool {
        return qtc.QFile_Open32(@ptrCast(self), @ptrCast(f), @intCast(ioFlags), @intCast(handleFlags));
    }

    pub fn Open33(self: ?*anyopaque, fd: i32, ioFlags: i32, handleFlags: i32) bool {
        return qtc.QFile_Open33(@ptrCast(self), fd, @intCast(ioFlags), @intCast(handleFlags));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QFile_Delete(@ptrCast(self));
    }
};

File: binding/src/libqfile.cpp
#include <QByteArray>
#include <QFile>
#include <QFileDevice>
#include <QNtfsPermissionCheckGuard>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qfile.h>
#include "libqfile.h"
#include "libqfile.hxx"

QNtfsPermissionCheckGuard* QNtfsPermissionCheckGuard_new() {
	 return new QNtfsPermissionCheckGuard();
}

void QNtfsPermissionCheckGuard_Delete(QNtfsPermissionCheckGuard* self) {
    delete self;
}

QFile* QFile_new() {
	 return new VirtualQFile();
}

QFile* QFile_new2(const libqt_string name) {
	 return new VirtualQFile(QString::fromUtf8(name.data, name.len));
}

QFile* QFile_new3(QObject* parent) {
	 return new VirtualQFile(parent);
}

QFile* QFile_new4(const libqt_string name, QObject* parent) {
	 return new VirtualQFile(QString::fromUtf8(name.data, name.len), parent);
}

libqt_string QFile_Tr(const char* s) {
	QString _ret = QFile::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFile_FileName(const QFile* self) {
	QString _ret = self->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QFile_SetFileName(QFile* self, const libqt_string name) {
	self->setFileName(QString::fromUtf8(name.data, name.len));
}

libqt_string QFile_EncodeName(const libqt_string fileName) {
	QByteArray _qb = QFile::encodeName(QString::fromUtf8(fileName.data, fileName.len));
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFile_DecodeName(const libqt_string localFileName) {
	QString _ret = QFile::decodeName(QByteArray(localFileName.data, localFileName.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFile_DecodeName2(const char* localFileName) {
	QString _ret = QFile::decodeName(localFileName);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QFile_Exists(const QFile* self) {
	return self->exists();
}

bool QFile_Exists2(const libqt_string fileName) {
	return QFile::exists(QString::fromUtf8(fileName.data, fileName.len));
}

libqt_string QFile_SymLinkTarget(const QFile* self) {
	QString _ret = self->symLinkTarget();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFile_SymLinkTarget2(const libqt_string fileName) {
	QString _ret = QFile::symLinkTarget(QString::fromUtf8(fileName.data, fileName.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QFile_Remove(QFile* self) {
	return self->remove();
}

bool QFile_Remove2(const libqt_string fileName) {
	return QFile::remove(QString::fromUtf8(fileName.data, fileName.len));
}

bool QFile_MoveToTrash(QFile* self) {
	return self->moveToTrash();
}

bool QFile_MoveToTrash2(const libqt_string fileName) {
	return QFile::moveToTrash(QString::fromUtf8(fileName.data, fileName.len));
}

bool QFile_Rename(QFile* self, const libqt_string newName) {
	return self->rename(QString::fromUtf8(newName.data, newName.len));
}

bool QFile_Rename2(const libqt_string oldName, const libqt_string newName) {
	return QFile::rename(QString::fromUtf8(oldName.data, oldName.len), QString::fromUtf8(newName.data, newName.len));
}

bool QFile_Link(QFile* self, const libqt_string newName) {
	return self->link(QString::fromUtf8(newName.data, newName.len));
}

bool QFile_Link2(const libqt_string fileName, const libqt_string newName) {
	return QFile::link(QString::fromUtf8(fileName.data, fileName.len), QString::fromUtf8(newName.data, newName.len));
}

bool QFile_Copy(QFile* self, const libqt_string newName) {
	return self->copy(QString::fromUtf8(newName.data, newName.len));
}

bool QFile_Copy2(const libqt_string fileName, const libqt_string newName) {
	return QFile::copy(QString::fromUtf8(fileName.data, fileName.len), QString::fromUtf8(newName.data, newName.len));
}

bool QFile_Open(QFile* self, int flags) {
	return self->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
}

bool QFile_Open2(QFile* self, int flags, int permissions) {
	return self->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags), static_cast<QFlags<QFileDevice::Permission>>(permissions));
}

bool QFile_Open4(QFile* self, int fd, int ioFlags) {
	return self->open(fd, static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(ioFlags));
}

long long QFile_Size(const QFile* self) {
	return self->size();
}

bool QFile_Resize(QFile* self, long long sz) {
	return self->resize(sz);
}

bool QFile_Resize2(const libqt_string filename, long long sz) {
	return QFile::resize(QString::fromUtf8(filename.data, filename.len), sz);
}

int QFile_Permissions(const QFile* self) {
	return self->permissions();
}

int QFile_Permissions2(const libqt_string filename) {
	return QFile::permissions(QString::fromUtf8(filename.data, filename.len));
}

bool QFile_SetPermissions(QFile* self, int permissionSpec) {
	return self->setPermissions(static_cast<QFlags<QFileDevice::Permission>>(permissionSpec));
}

bool QFile_SetPermissions2(const libqt_string filename, int permissionSpec) {
	return QFile::setPermissions(QString::fromUtf8(filename.data, filename.len), static_cast<QFlags<QFileDevice::Permission>>(permissionSpec));
}

libqt_string QFile_Tr2(const char* s, const char* c) {
	QString _ret = QFile::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFile_Tr3(const char* s, const char* c, int n) {
	QString _ret = QFile::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QFile_MoveToTrash22(const libqt_string fileName, libqt_string pathInTrash) {
	return QFile::moveToTrash(QString::fromUtf8(fileName.data, fileName.len), QString::fromUtf8(pathInTrash.data, pathInTrash.len));
}

bool QFile_Open33(QFile* self, int fd, int ioFlags, int handleFlags) {
	return self->open(fd, static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(ioFlags), static_cast<QFlags<QFileDevice::FileHandleFlag>>(handleFlags));
}

// Base class handler implementation
libqt_string QFile_QBaseFileName(const QFile* self) {
	auto* vqfile = dynamic_cast<const VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_FileName_IsBase(true);
	QString _ret = vqfile->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QFile_OnFileName(const QFile* self, intptr_t slot) {
	auto* vqfile = dynamic_cast<const VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_FileName_Callback(reinterpret_cast<VirtualQFile::QFile_FileName_Callback>(slot));
}
}

// Base class handler implementation
bool QFile_QBaseOpen(QFile* self, int flags) {
	auto* vqfile = dynamic_cast<VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Open_IsBase(true);
	return vqfile->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
}
}

// Auxiliary method to allow providing re-implementation
void QFile_OnOpen(QFile* self, intptr_t slot) {
	auto* vqfile = dynamic_cast<VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Open_Callback(reinterpret_cast<VirtualQFile::QFile_Open_Callback>(slot));
}
}

// Base class handler implementation
long long QFile_QBaseSize(const QFile* self) {
	auto* vqfile = dynamic_cast<const VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Size_IsBase(true);
	return vqfile->size();
}
}

// Auxiliary method to allow providing re-implementation
void QFile_OnSize(const QFile* self, intptr_t slot) {
	auto* vqfile = dynamic_cast<const VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Size_Callback(reinterpret_cast<VirtualQFile::QFile_Size_Callback>(slot));
}
}

// Base class handler implementation
bool QFile_QBaseResize(QFile* self, long long sz) {
	auto* vqfile = dynamic_cast<VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Resize_IsBase(true);
	return vqfile->resize(sz);
}
}

// Auxiliary method to allow providing re-implementation
void QFile_OnResize(QFile* self, intptr_t slot) {
	auto* vqfile = dynamic_cast<VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Resize_Callback(reinterpret_cast<VirtualQFile::QFile_Resize_Callback>(slot));
}
}

// Base class handler implementation
int QFile_QBasePermissions(const QFile* self) {
	auto* vqfile = dynamic_cast<const VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Permissions_IsBase(true);
	return vqfile->permissions();
}
}

// Auxiliary method to allow providing re-implementation
void QFile_OnPermissions(const QFile* self, intptr_t slot) {
	auto* vqfile = dynamic_cast<const VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_Permissions_Callback(reinterpret_cast<VirtualQFile::QFile_Permissions_Callback>(slot));
}
}

// Base class handler implementation
bool QFile_QBaseSetPermissions(QFile* self, int permissionSpec) {
	auto* vqfile = dynamic_cast<VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_SetPermissions_IsBase(true);
	return vqfile->setPermissions(static_cast<QFlags<QFileDevice::Permission>>(permissionSpec));
}
}

// Auxiliary method to allow providing re-implementation
void QFile_OnSetPermissions(QFile* self, intptr_t slot) {
	auto* vqfile = dynamic_cast<VirtualQFile*>(self);
	if (vqfile && vqfile->isVirtualQFile) {
vqfile->setQFile_SetPermissions_Callback(reinterpret_cast<VirtualQFile::QFile_SetPermissions_Callback>(slot));
}
}

void QFile_Delete(QFile* self) {
    delete self;
}

